### DISCUSSION TITLE
Cannot register commands with cogs

### DISCUSSION THREAD

**Discussion by aurkaxi** [Rapptz/discord.py#8380] [Q&A] [answered]
```
import discord, os
from discord.ext import commands


bot = discord.ext.commands.Bot(command_prefix="&", intents=discord.Intents.all())

class Test(commands.Cog):
	def __init__(self, bot):
		self.bot = bot

	@commands.command()
	async def ping(self, ctx, name: str = None):
		await ctx.send("Pong")

bot.add_cog(Test(bot))

token = os.environ['TOKEN']
bot.run(token)
```
while i send message "&ping" nothing happens. on the console the log is 
```
ERROR    discord.ext.commands.bot Ignoring exception in command None
discord.ext.commands.errors.CommandNotFound: Command "ping" is not found
```

In this [video][1] it works completely fine.

  [1]: https://youtu.be/g1xESivzCPE?list=PLESMQx4LeD3NmTZ8D1qwQwwSp67kznl-K&t=246

**Comment by izxxr** [comment, marked as answer]
As of discord.py v2.0, `Bot.add_cog` is a coroutine function. You must `await` it. For detailed information on asynchronous setup, See [this section](https://discordpy.readthedocs.io/en/latest/migrating.html#extension-and-cog-loading-unloading-is-now-asynchronous) on migrating guide.

**Reply by aurkaxi** [reply]
So i made a folder 'cogs' created two files inside them. one in '__init__.py' which is blank. another one is test.py with 
```
from discord.ext import commands
class Test(commands.Cog):
	def __init__(self, bot):
		self.bot = bot

	@commands.command()
	async def ping(self, ctx, name: str = None):
		await ctx.send("Pong")

async def setup(bot):
	await bot.add_cog(Test(bot))
```

and my main.py is 
```
import discord, os
from discord.ext import commands
from keep_alive import keep_alive

bot = discord.ext.commands.Bot(command_prefix="&", intents=discord.Intents.all())

async def setup_hook():
	await bot.load_extension(f'cogs.test')

token = os.environ['TOKEN']
bot.run(token)
```


i still can't figure out.

**Reply by bitterteriyaki** [reply]
You have to replace the default setup hook. You are defining a new hook but never replacing the old one. You can do this using something like this:
```py
import discord, os
from discord.ext import commands
from keep_alive import keep_alive

bot = discord.ext.commands.Bot(command_prefix="&", intents=discord.Intents.all())

async def setup_hook():
	await bot.load_extension(f'cogs.test')

bot.setup_hook = setup_hook

token = os.environ['TOKEN']
bot.run(token)
```

**Reply by izxxr** [reply]
`setup_hook` is a method of `Bot` class that you have to override. You have to subclass `Bot` and override that method.
```py
class Bot(commands.Bot):
  async def setup_hook(self):
      ...

bot = Bot(...)
```
If you don't want to subclass the `Bot` class, you can use an alternative method for async setup by using `asyncio.run()` with your own async function used as entry point. An example is given in the section I mentioned above.
